src/include/distributed/multi_physical_planner.h

```diff
@@ -203,42 +203,64 @@ typedef struct MapMergeJob
  */
 typedef struct TaskExecution TaskExecution;
 
+typedef enum TaskQueryType
+{
+	TASK_QUERY_TEXT,
+	TASK_QUERY_OBJECT,
+	TASK_QUERY_TEXT_PER_PLACEMENT
+} TaskQueryType;
+
+typedef struct TaskQuery
+{
+	CitusNode type;
+	TaskQueryType queryType;
+
+	union
+	{
+		/*
+		 * For most queries queryForLocalExecution and/or queryStringLazy is not
+		 * NULL. This means we have a single query for all placements.
+		 *
+		 * If this is not the case, the length of perPlacementQueryStrings is
+		 * non-zero and equal to length of taskPlacementList. Like this it can
+		 * assign a different query for each placement. We need this flexibility
+		 * when a query should return node specific values. For example, on which
+		 * node did we succeed storing some result files?
+		 *
+		 * queryForLocalExecution is only not null when the planner thinks the
+		 * query could possibly be locally executed. In that case deparsing+parsing
+		 * the query might not be necessary, so we do that lazily.
+		 *
+		 * queryForLocalExecution should only be set by using SetTaskQueryIfShouldLazyDeparse()
+		 */
+		Query *queryForLocalExecution;
+
+		/*
+		 * In almost all cases queryStringLazy should be read only indirectly by
+		 * using TaskQueryStringAllPlacements(). This will populate the field if only the
+		 * queryForLocalExecution field is not NULL.
+		 *
+		 * This field should only be set by using SetTaskQueryString() (or as a
+		 * side effect from TaskQueryStringAllPlacements()). Otherwise it might not be in sync
+		 * with queryForLocalExecution.
+		 */
+		char *queryStringLazy;
```

Comment:
Putting this in a union doesn't seem right. There are cases in which we have both queryForLocalExecution and queryStringLazy (e.g. write to a reference table with a placement on the coordinator), so if we set one we override the other.

Reply:
Talked offline and we realized that queryForLocalExecution wasn't actually used other than lazy deparsing. Therefore it is renamed as `jobQueryReferenceForLazyDeparsing`.

Also I have added queryStringList into this union.